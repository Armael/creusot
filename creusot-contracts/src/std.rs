pub mod clone;
pub mod cmp;
pub mod default;
pub mod eq;
mod fun;
pub mod mem;
pub mod option;
mod slice;
pub mod vec;

pub use clone::*;
pub use eq::*;
pub use fun::*;
pub use vec::*;
