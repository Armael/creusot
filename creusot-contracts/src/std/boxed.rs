use crate as creusot_contracts;
use creusot_contracts_proc::*;

use std::alloc::Allocator;

extern_spec! {
    mod std {
        mod boxed {
            impl<T, A: Allocator> Box<T> {
                #[ensures(**self == *result)]
                #[ensures(*^self == ^result)]
                fn as_mut(&mut self) -> &mut T;

                #[ensures(**self == *result)]
                fn as_ref(&self) -> &T;
            }
        }
    }
}
