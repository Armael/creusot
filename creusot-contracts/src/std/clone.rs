use crate::macros::*;

extern_spec! {
    mod std {
        mod clone {
            trait Clone {
                // TODO:
                // Requiring result == *self seems too strong since objects can
                // contain information which is lost by the clone (e.g., capacity of
                // vectors...).
                #[ensures(result == *self)]
                fn clone(&self) -> Self;
            }
        }
    }
}
