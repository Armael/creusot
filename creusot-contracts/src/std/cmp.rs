use crate as creusot_contracts;
use creusot_contracts_proc::*;

use crate::logic::ord::*;
pub use std::cmp::Ordering;

pub trait Ord: OrdLogic + Eq {
    #[ensures(result == self.cmp_log(*o))]
    fn cmp(&self, o: &Self) -> Ordering;

    #[ensures(result == (*self <= *o))]
    fn le(&self, o: &Self) -> bool {
        match self.cmp(o) {
            Ordering::Greater => false,
            _ => true,
        }
    }

    #[ensures(result == (*self >= *o))]
    fn ge(&self, o: &Self) -> bool {
        match self.cmp(o) {
            Ordering::Less => false,
            _ => true,
        }
    }

    #[ensures(result == (*self > *o))]
    fn gt(&self, o: &Self) -> bool {
        match self.cmp(o) {
            Ordering::Greater => true,
            _ => false,
        }
    }

    #[ensures(result == (*self < *o))]
    fn lt(&self, o: &Self) -> bool {
        match self.cmp(o) {
            Ordering::Less => true,
            _ => false,
        }
    }
}

// The new spec, but peacefully coexisting with the previous approach
mod real_ord {
    use crate as creusot_contracts;
    use crate::{logic::ord::*, Model};
    use creusot_contracts_proc::*;
    pub use std::cmp::Ordering;

    extern_spec! {
        mod std {
            mod cmp {
                trait Ord
                where Self: Model,
                      Self::ModelTy: OrdLogic {

                    #[ensures(result == (@self).cmp_log(@*rhs))]
                    fn cmp(&self, rhs: &Self) -> Ordering;

                    #[ensures(@result >= @self)]
                    #[ensures(@result >= @o)]
                    #[ensures(result == self || result == o)]
                    #[ensures(@self <= @o ==> result == o)]
                    #[ensures(@o < @self ==> result == self)]
                    fn max(self, o: Self_) -> Self;
                }
            }
        }
    }
}
