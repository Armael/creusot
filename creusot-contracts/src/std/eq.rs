use crate as creusot_contracts;
use creusot_contracts_proc::*;

use crate::logic::*;

extern_spec! {
    #[ensures(result === (@self_ === @rhs))]
    fn std::cmp::PartialEq::eq<Self_, Rhs>(self_: &Self_, rhs: &Rhs) -> bool
        where Self_ : PartialEq<Rhs>,
              Self_ : Model,
              Rhs: Model<ModelTy = Self_::ModelTy>,
}
