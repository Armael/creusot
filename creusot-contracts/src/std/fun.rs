use crate as creusot_contracts;
use creusot_contracts_proc::*;

#[rustc_diagnostic_item = "fn_once_spec"]
pub trait FnOnceSpec<Args>: FnOnce<Args> {
    #[predicate]
    fn precondition(self, a: Args) -> bool;

    #[predicate]
    fn postcondition_once(self, a: Args, res: Self::Output) -> bool;
}

#[rustc_diagnostic_item = "fn_mut_spec"]
pub trait FnMutSpec<Args>: FnMut<Args> + FnOnceSpec<Args> {
    #[predicate]
    fn postcondition_mut(&mut self, a: Args, res: Self::Output) -> bool;

    #[law]
    #[ensures((exists<s: &mut Self> *s === self && s.postcondition_mut(a, res) && (^s).resolve()) ==> self.postcondition_once(a, res))]
    #[ensures(self.postcondition_once(a, res) ==> exists<s: &mut Self> *s === self && s.postcondition_mut(a, res) && (^s).resolve())]
    fn fn_mut_once(self, a: Args, res: Self::Output)
    where
        Self: crate::Resolve + Sized;
}

#[rustc_diagnostic_item = "fn_spec"]
pub trait FnSpec<Args>: Fn<Args> + FnMutSpec<Args> {
    #[predicate]
    fn postcondition(&self, _: Args, _: Self::Output) -> bool;

    #[law]
    #[ensures(self.postcondition(args, res) ==> exists<s: &mut Self> *s === *self && s.resolve() && s.postcondition_mut(args, res))]
    #[ensures((exists<s: &mut Self> *s === *self && s.resolve() && s.postcondition_mut(args, res)) ==> self.postcondition(args, res))]
    fn fn_mut(&self, args: Args, res: Self::Output)
    where
        Self: crate::Resolve + Sized;
}

impl<Args, F: FnOnce<Args>> FnOnceSpec<Args> for F {
    #[predicate]
    #[trusted]
    #[rustc_diagnostic_item = "fn_once_impl_precond"]
    fn precondition(self, _: Args) -> bool {
        std::process::abort()
    }

    #[predicate]
    #[trusted]
    #[rustc_diagnostic_item = "fn_once_impl_postcond"]
    fn postcondition_once(self, _: Args, _: Self::Output) -> bool {
        std::process::abort()
    }
}

impl<Args, F: FnMut<Args>> FnMutSpec<Args> for F {
    #[predicate]
    #[trusted]
    #[rustc_diagnostic_item = "fn_mut_impl_postcond"]
    fn postcondition_mut(&mut self, _: Args, _: Self::Output) -> bool {
        std::process::abort()
    }

    #[law]
    fn fn_mut_once(self, _: Args, _: Self::Output) {}
}

impl<Args, F: Fn<Args>> FnSpec<Args> for F {
    #[predicate]
    #[trusted]
    #[rustc_diagnostic_item = "fn_impl_postcond"]
    fn postcondition(&self, _: Args, _: Self::Output) -> bool {
        std::process::abort()
    }

    #[law]
    fn fn_mut(&self, _: Args, _: Self::Output) {}
}

extern_spec! {
    #[requires(f.precondition(a))]
    #[ensures(f.postcondition_once(a, result))]
    fn std::ops::FnOnce::call_once<A, F : FnOnceSpec<A>>(f: F, a: A) -> F::Output
}

extern_spec! {
    #[requires((*f).precondition(a))]
    #[ensures(f.postcondition_mut(a, result))]
    fn std::ops::FnMut::call_mut<A, F : FnMutSpec<A>>(f: &mut F, a: A) -> F::Output
}

extern_spec! {
    #[requires((*f).precondition(a))]
    #[ensures(f.postcondition(a, result))]
    fn std::ops::Fn::call<A, F : FnSpec<A>>(f: &F, a: A) -> F::Output
}
