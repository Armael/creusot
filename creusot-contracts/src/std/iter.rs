use crate as creusot_contracts;
use crate::{Int, Model, OrdLogic, Seq};
use creusot_contracts_proc::*;
use std::iter::Skip;

pub trait IteratorSpec: Iterator {
    #[predicate]
    fn produces(self, visited: Seq<Self::Item>, _: Self) -> bool;

    #[predicate]
    fn completed(&mut self) -> bool;

    #[law]
    #[ensures(a.produces(Seq::EMPTY, a))]
    fn produces_refl(a: Self);

    #[law]
    #[requires(a.produces(ab, b))]
    #[requires(b.produces(bc, c))]
    #[ensures(a.produces(ab.concat(bc), c))]
    fn produces_trans(a: Self, ab: Seq<Self::Item>, b: Self, bc: Seq<Self::Item>, c: Self);
}

extern_spec! {
    mod std {
        mod iter {
            trait Iterator
                where Self : IteratorSpec {

                #[ensures(match result {
                    None => self.completed(),
                    Some(v) => (*self).produces(Seq::singleton(v), ^self)
                })]
                fn next(&mut self) -> Option<Self_::Item>;

                #[ensures(@result == (self, @n))]
                fn skip(self, n: usize) -> Skip<Self_>;
            }
        }
    }
}

extern_spec! {
    impl<I : Iterator> IntoIterator for I {
        #[ensures(result == self)]
        fn into_iter(self) -> I;
    }
}

impl<I> Model for Skip<I> {
    type ModelTy = (I, Int);

    #[logic]
    #[trusted]
    fn model(self) -> Self::ModelTy {
        pearlite! { absurd }
    }
}

impl<I: IteratorSpec> IteratorSpec for Skip<I> {
    #[predicate]
    fn completed(&mut self) -> bool {
        pearlite! {
            (@^self).1 == 0 &&
            exists<s: Seq<Self::Item>, i: &mut I>
                s.len() <= (@*self).1 &&
                (@self).0.produces(s, *i) &&
                i.completed() &&
                ^i == (@^self).0
        }
    }

    #[predicate]
    fn produces(self, visited: Seq<Self::Item>, o: Self) -> bool {
        pearlite! {
            visited == Seq::EMPTY && self == o ||
            (@o).1 == 0 &&
            exists<s: Seq<Self::Item>>
                s.len() == (@self).1 &&
                (@self).0.produces(s.concat(visited), (@o).0) &&
                (s.len() < (@self).1 ==> visited == Seq::EMPTY)
        }
    }

    #[law]
    #[ensures(a.produces(Seq::EMPTY, a))]
    fn produces_refl(a: Self) {}

    #[law]
    #[requires(a.produces(ab, b))]
    #[requires(b.produces(bc, c))]
    #[ensures(a.produces(ab.concat(bc), c))]
    fn produces_trans(a: Self, ab: Seq<Self::Item>, b: Self, bc: Seq<Self::Item>, c: Self) {}
}
