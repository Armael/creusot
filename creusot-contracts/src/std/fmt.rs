use crate::*;
#[cfg(feature = "contracts")]
use ::std::fmt::{ArgumentV1, Arguments, Debug, Formatter};

extern_spec! {
    mod std {
        mod io {
            #[requires(true)]
            fn _print(fmt: std::fmt::Arguments<'_>);
        }

        mod fmt {
            impl<'a> ArgumentV1<'a> {
            #[requires(true)]
            fn new_debug<'b, T : Debug>(x: &'b T) -> ArgumentV1<'_>;
            }

            impl<'a> Arguments<'a> {
                #[requires(true)]
                fn new_v1(pieces: &'a [&'static str], args: &'a [std::fmt::ArgumentV1<'a>]) -> std::fmt::Arguments<'a>;
            }

            impl<'a> Formatter<'a> {
                #[requires(true)]
                fn debug_struct_field1_finish<'b>(
                    &'b mut self,
                    name: &str,
                    name1: &str,
                    value1: &dyn Debug,
                ) -> ::std::fmt::Result;
            }
        }
    }
}
