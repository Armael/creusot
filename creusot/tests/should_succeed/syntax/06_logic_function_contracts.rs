// WHY3PROVE Z3
extern crate creusot_contracts;

use creusot_contracts::std::*;
use creusot_contracts::*;

#[logic]
#[variant(seq.len())]
fn sum(seq: Seq<Int>) -> Int {
    pearlite! {
        if seq.len() === 0 { 0 }
        else {
            seq[seq.len() - 1] + sum(seq.subsequence(0, seq.len() - 1))
        }
    }
}
