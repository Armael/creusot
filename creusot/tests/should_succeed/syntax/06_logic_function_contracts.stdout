module Type
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
end
module C06LogicFunctionContracts_Sum_Interface
  use seq.Seq
  use mach.int.Int
  function sum (seq : Seq.seq int) : int
end
module C06LogicFunctionContracts_Sum
  use seq.Seq
  use mach.int.Int
  use mach.int.Int32
  use seq_ext.SeqExt
  function sum (seq : Seq.seq int) : int
  axiom def : forall seq : Seq.seq int . sum seq = (if Seq.length seq = 0 then
    0
  else
    Seq.get seq (Seq.length seq - 1) + sum (SeqExt.subsequence seq 0 (Seq.length seq - 1))
  )
  axiom sum_spec : forall seq : Seq.seq int . true
end
module C06LogicFunctionContracts_Sum_Impl
  use seq.Seq
  use mach.int.Int
  use mach.int.Int32
  use seq_ext.SeqExt
  let rec ghost function sum (seq : Seq.seq int) : int
    variant {Seq.length seq}
    
   = 
    if Seq.length seq = 0 then
      0
    else
      Seq.get seq (Seq.length seq - 1) + sum (SeqExt.subsequence seq 0 (Seq.length seq - 1))
    
end
module C06LogicFunctionContracts_AllZero_Interface
  use seq.Seq
  use mach.int.Int
  predicate all_zero (seq : Seq.seq int)
end
module C06LogicFunctionContracts_AllZero
  use seq.Seq
  use mach.int.Int
  use mach.int.Int32
  use seq_ext.SeqExt
  predicate all_zero (seq : Seq.seq int)
  axiom def : forall seq : Seq.seq int . all_zero seq = (if Seq.length seq = 0 then
    true
  else
    Seq.get seq (Seq.length seq - 1) = 0 && all_zero (SeqExt.subsequence seq 0 (Seq.length seq - 1))
  )
  axiom all_zero_spec : forall seq : Seq.seq int . true
end
module C06LogicFunctionContracts_AllZero_Impl
  use seq.Seq
  use mach.int.Int
  use mach.int.Int32
  use seq_ext.SeqExt
  let rec ghost function all_zero (seq : Seq.seq int)
    variant {Seq.length seq}
    
   = 
    if Seq.length seq = 0 then
      true
    else
      Seq.get seq (Seq.length seq - 1) = 0 && all_zero (SeqExt.subsequence seq 0 (Seq.length seq - 1))
    
end
